### PR TITLE
Refactor jail to handle ssh_* methods specially

### PR DESCRIPTION
**Pull request by influx6**
PR refactors jail to handle specific RPC request for `ssh_*` based methods using the local node's `rpc.Client`. This ensure whispers rpc calls are handle  internally.

Related to issue #274.

### CODE REVIEW

File: geth/common/types.go [hunk 1]
```diff
@@ -88,6 +88,12 @@ type NodeManager interface {
 	// RPCClient exposes reference to RPC client connected to the running node
```
Comment (adambabik) [inline geth/common/types.go]: We should think about removing this method then.

File: geth/jail/execution_policy.go [hunk 2]
```diff
@@ -142,6 +142,72 @@ func (ep *ExecutionPolicy) executeLocalSendTransaction(req common.RPCCall, call
 	return resp, nil
 }
 
+// ExecuteSHH defines a function which handles the processing of `shh_*` transaction methods
+// rpc request to the internal node server.
+func (ep *ExecutionPolicy) ExecuteSHH(req common.RPCCall, call otto.FunctionCall) (*otto.Object, error) {
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: This looks like a copy paste of `ExecuteOtherTransaction()`. Maybe we could somehow combine them?

File: geth/jail/jail.go [hunk 3]
```diff
@@ -204,8 +205,10 @@ func (jail *Jail) Send(call otto.FunctionCall) (response otto.Value) {
 		var resErr error
 		var res *otto.Object
 
-		switch req.Method {
-		case params.SendTransactionMethodName:
+		switch {
```
Comment (adambabik) [inline geth/jail/jail.go]: This solution might not scale as we would need to create new methods like `ExecuteXXX()` with every exception. It looks like we need:
* `policy.ExecuteSendTransaction()` as it's unique, 
* `policy.ExecuteLocally()` as some commands can only be executed locally,
* `policy.Execute???()` as some can be executed using both clients.

File: geth/node/manager.go [hunk 4]
```diff
@@ -550,6 +575,25 @@ func (m *NodeManager) RPCClient() (*rpc.Client, error) {
 	return m.rpcClient, nil
 }
 
+// RPCClient exposes reference to RPC client connected to the running node.
+func (m *NodeManager) RPCClient() (*rpc.Client, error) {
```
Comment (adambabik) [inline geth/node/manager.go]: As I mentioned above, I am not sure we need this one. It might introduce some confusion which method use and can also result in bugs that will be hard to figure out. Caller should be 100% sure which client it needs to use.

File: geth/common/types.go (hunk 1 above)
Comment (influx6) [reply geth/common/types.go]: Yes, but first we must find out which areas use this and their context, to understand how the change affects the code.

Comment (adambabik) [reply geth/common/types.go]: Sure. It can be done even later as part of some refactoring.

File: geth/node/manager.go (hunk 4 above)
Comment (influx6) [reply geth/node/manager.go]: Am in agreement with this, but lets push it into a refactor issue, where we can deal with the implications of removal.

File: geth/jail/execution_policy.go [hunk 5]
```diff
@@ -30,18 +32,58 @@ func NewExecutionPolicy(nodeManager common.NodeManager, accountManager common.Ac
...
-	if config.UpstreamConfig.Enabled {
-		return ep.executeRemoteSendTransaction(req, call)
+	var res *otto.Object
+	var resErr error
+
+	switch {
+	case strings.HasPrefix(req.Method, "shh_"):
+		res, resErr = ep.ExecuteLocally(req, call)
+	case params.SendTransactionMethodName == req.Method:
+		switch config.UpstreamConfig.Enabled {
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: Simple `if` statement would allow to avoid nested `switch`es.

File: geth/jail/execution_policy.go [hunk 6]
```diff
@@ -30,18 +32,58 @@ func NewExecutionPolicy(nodeManager common.NodeManager, accountManager common.Ac
...
+		case true:
+			res, resErr = ep.executeRemoteSendTransaction(req, call)
+		case false:
+			res, resErr = ep.executeLocalSendTransaction(req, call)
+		}
+	default:
+		switch config.UpstreamConfig.Enabled {
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: The same here.

File: geth/jail/execution_policy.go [hunk 7]
```diff
@@ -30,18 +32,58 @@ func NewExecutionPolicy(nodeManager common.NodeManager, accountManager common.Ac
...
+	var res *otto.Object
+	var resErr error
+
+	switch {
+	case strings.HasPrefix(req.Method, "shh_"):
+		res, resErr = ep.ExecuteLocally(req, call)
+	case params.SendTransactionMethodName == req.Method:
+		switch config.UpstreamConfig.Enabled {
+		case true:
+			res, resErr = ep.executeRemoteSendTransaction(req, call)
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: Any reason why `execute*SendTransaction` methods are not exported while `ExecuteLocally` and `ExecuteOnRemote` are exported?

File: geth/jail/execution_policy.go [hunk 8]
```diff
@@ -30,18 +32,58 @@ func NewExecutionPolicy(nodeManager common.NodeManager, accountManager common.Ac
...
+// ExecuteLocally defines a function which handles the processing of `shh_*` transaction methods
+// rpc request to the internal node server.
+func (ep *ExecutionPolicy) ExecuteLocally(req common.RPCCall, call otto.FunctionCall) (*otto.Object, error) {
+	client, err := ep.nodeManager.RPCLocalClient()
+	if err != nil {
+		return nil, common.StopRPCCallError{Err: err}
 	}
 
-	return ep.executeLocalSendTransaction(req, call)
+	return ep.executedWithClient(client, req, call)
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: Typo? Shouldn't it be `executeWithClient()`?

File: geth/node/manager.go [hunk 9]
```diff
@@ -502,8 +502,40 @@ func (m *NodeManager) AccountKeyStore() (*keystore.KeyStore, error) {
...
+	if m.rpcClient == nil {
+		var err error
+		m.rpcClient, err = m.node.Attach()
+		if err != nil {
+			log.Error("Cannot attach RPC client to node", "error", err)
+			return nil, ErrInvalidRPCClient
+		}
+	}
+
+	if m.rpcClient == nil {
```
Comment (adambabik) [inline geth/node/manager.go]: I guess it's not possible as the same condition is checked a few lines above.

File: geth/node/manager.go [hunk 10]
```diff
@@ -523,24 +555,12 @@ func (m *NodeManager) RPCClient() (*rpc.Client, error) {
...
-	if m.rpcClient == nil {
-		var err error
-		m.rpcClient, err = m.node.Attach()
-		if err != nil {
-			log.Error("Cannot attach RPC client to node", "error", err)
-			return nil, ErrInvalidRPCClient
-		}
 	}
 
 	if m.rpcClient == nil {
```
Comment (adambabik) [inline geth/node/manager.go]: The same here. `if m.rpcClient == nil {` is checked a few lines above.

File: geth/jail/execution_policy.go (hunk 7 above)
Comment (influx6) [reply geth/jail/execution_policy.go]: `execute*sendTransaction` are unexported, has I felt a single `ExecuteSendTransaction` would be a better encapsulation for the different needed implementation for upstream/local.

**Comment by influx6** [comment]
Attached is google docs spreadsheet for JSON RPC commands routing:
https://docs.google.com/spreadsheets/d/1N1nuzVN5tXoDmzkBLeC9_mwIlVH8DGF7YD2XwxA8BAE/edit?usp=sharing

File: geth/jail/execution_policy.go (hunk 7 above)
Comment (influx6) [reply geth/jail/execution_policy.go]: `execute*sendTransaction` are unexported, has I felt a single `ExecuteSendTransaction` would be a better encapsulation for the different needed implementation for upstream/local.

File: geth/jail/execution_policy.go [hunk 11]
```diff
@@ -13,6 +13,89 @@ import (
...
+
+// local and upstream RouteID
+const (
+	localCommand rpcRouteID = iota + 1
+	upstreamCommand
+)
+
+// map of command routes
+var (
+	rpcCommandsRoute = map[string]rpcRouteID{
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: Do we want to maintain this map? If a new Whisper command is added, we need to add it here. We could keep only a list of exceptions from the rules instead.

File: geth/jail/execution_policy.go [hunk 12]
```diff
@@ -30,22 +113,62 @@ func NewExecutionPolicy(nodeManager common.NodeManager, accountManager common.Ac
...
+		if config.UpstreamConfig.Enabled {
+			return ep.ExecuteRemoteSendTransaction(req, call)
+		}
+
+		return ep.ExecuteLocalSendTransaction(req, call)
+	}
+
+	var res *otto.Object
+	var resErr error
+	switch rpcCommandsRoute[req.Method] {
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: 1. No `default` case.
2. Each line could `return ...` immediately instead of having `var`s as nothing is done with these variables anyway.

File: geth/jail/execution_policy.go [hunk 13]
```diff
@@ -13,6 +13,43 @@ import (
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/robertkrimen/otto"
 	"github.com/status-im/status-go/geth/common"
+	"github.com/status-im/status-go/geth/params"
+)
+
+// map of command routes
+var (
+	rpcLocalCommandRoute = map[string]bool{
```
Comment (adambabik) [inline geth/jail/execution_policy.go]: It's not exactly what I meant but I also think it's good enough to be merged. 

I meant to remove the map completely and rely on a function (functions) that maps a command name to the execution strategy. Here, if a new Whisper command is added to the protocol, we need to change the code and release a new version of the app. 

This PR is in the queue for a long time so if you could just add a comment with `TODO`, it would be sufficient.

Comment (influx6) [reply geth/jail/execution_policy.go]: Hmmm, I can quick run down up, but that means we would have a map of commands to functions. We can shift it to another issue then.

Comment (adambabik) [reply geth/jail/execution_policy.go]: It depends. So, for instance, Whisper commands could be detected with a function:

```go
func (cmd string) (Route, bool) {
  if strings.HasPrefix("shh_", cmd) {
    return LocalRoute, true
  }

  return UnknownRoute, false
}
```

For specific `eth_` commands you would need to still rely on a map but encapsulated within a function with the same signature:
```go
func (cmd string) (Route, bool) {
  return localEthRoutes[cmd]
}
```

Finally, if there is no match, execute remotely.

File: geth/jail/execution_policy.go [hunk 14]
```diff
@@ -30,22 +69,56 @@ func NewExecutionPolicy(nodeManager common.NodeManager, accountManager common.Ac
...
+		return ep.ExecuteLocalSendTransaction(req, call)
+	}
+
+	return ep.ExecuteLocally(req, call)
 }
 
-// executeRemoteSendTransaction defines a function to execute RPC method eth_sendTransaction over the upstream server.
-func (ep *ExecutionPolicy) executeRemoteSendTransaction(req common.RPCCall, call otto.FunctionCall) (*otto.Object, error) {
+// ExecuteLocally defines a function which handles the processing of `shh_*` transaction methods
+// rpc request to the internal node server.
```
Comment (tiabc) [inline geth/jail/execution_policy.go]: The comment is lying.

File: geth/node/manager.go [hunk 15]
```diff
@@ -502,8 +502,36 @@ func (m *NodeManager) AccountKeyStore() (*keystore.KeyStore, error) {
...
+
+	m.RLock()
+	defer m.RUnlock()
+
+	// make sure that node is fully started
+	if m.node == nil || m.nodeStarted == nil {
+		return nil, ErrNoRunningNode
+	}
+
+	<-m.nodeStarted
```
Comment (tiabc) [inline geth/node/manager.go]: Are you sure we need all these checks?

File: geth/jail/execution_policy.go [hunk 16]
```diff
@@ -13,6 +13,45 @@ import (
...
+		"net_listening": true,
+
+		// blockchain commands
+		"eth_sign":            true,
+		"eth_accounts":        true,
+		"eth_getCompilers":    true,
+		"eth_compileLLL":      true,
+		"eth_compileSolidity": true,
+		"eth_compileSerpent":  true,
+	}
```
[the remark this answers is not in the document]
Comment (adambabik) [reply geth/jail/execution_policy.go]: Re 1: I am not sure you have seen that but we talked about this idea https://github.com/status-im/status-go/pull/276#discussion_r136606210 as one possible way of refactoring the map. The idea to shift from selecting which commands run locally to which commands run on upstream is definitely good.

[the remark this answers is not in the document]
Comment (tiabc) [reply geth/jail/execution_policy.go]: That's not exactly it. You're talking about mapping requests to execution policies at compile time while I'm about constructing appropriate objects. But this is good that this question arose at your discussion, anyway, I've seen it.

File: geth/node/manager.go (hunk 15 above)
Comment (influx6) [reply geth/node/manager.go]: In my opinion, the `m.nodeStarted` is abit overkill but the checks are necessary even here.

File: geth/jail/execution_policy.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (influx6) [reply geth/jail/execution_policy.go]: @tiabc In effort to understand your thoughts, here is what i imagine:

```
type Policy interface{
  CanExecute(RPCCall) bool
  Execute(RPCCall, otto.FunctionCall)
}

newPolicy := ExecutionPolicies(policies ...Policy)

```

`newPolicy` can then be passed to `JailManager` and any other structures as needed.

Hope this embodies what you envisioned?

File: geth/node/txqueue_manager.go [hunk 17]
```diff
@@ -162,8 +162,10 @@ func (m *TxQueueManager) CompleteTransaction(id common.QueuedTxID, password stri
 	var txErr error
 
 	if config.UpstreamConfig.Enabled {
+		log.Info("completing completed upstream", "id", queuedTx.ID)
```
Comment (adambabik) [inline geth/node/txqueue_manager.go]: There is a log inside `completeRemoteTransaction`.

File: geth/jail/jail_rpc_test.go [hunk 18]
```diff
@@ -147,6 +156,13 @@ func (s *JailRPCTestSuite) TestSendTransaction() {
 	selectErr := s.Account.SelectAccount(TestConfig.Account1.Address, TestConfig.Account1.Password)
 	require.NoError(selectErr)
 
+	// TODO(influx6): How do we move this into ExecutionPolicy.executeSendTransaction since
+	// we also need the account's password.
+	// TransactionQueueHandler is required to enqueue a transaction.
+	s.TxQueueManager.SetTransactionQueueHandler(func(queuedTx *common.QueuedTx) {
+		s.TxQueueManager.CompleteTransaction(queuedTx.ID, TestConfig.Account1.Password)
```
Comment (adambabik) [inline geth/jail/jail_rpc_test.go]: Let's check if returned error is `nil`.

File: geth/jail/jail_rpc_test.go [hunk 19]
```diff
@@ -147,6 +156,13 @@ func (s *JailRPCTestSuite) TestSendTransaction() {
 	selectErr := s.Account.SelectAccount(TestConfig.Account1.Address, TestConfig.Account1.Password)
 	require.NoError(selectErr)
 
+	// TODO(influx6): How do we move this into ExecutionPolicy.executeSendTransaction since
```
Comment (adambabik) [inline geth/jail/jail_rpc_test.go]: I don't really get this comment. Could you elaborate? What would like to move into `ExecutionPolicy.executeSendTransaction`?

File: geth/jail/jail_rpc_test.go [hunk 20]
```diff
@@ -42,27 +43,35 @@ func TestJailRPCTestSuite(t *testing.T) {
...
+	Account        common.AccountManager
+	Policy         *jail.ExecutionPolicy
 }
 
 func (s *JailRPCTestSuite) SetupTest() {
 	require := s.Require()
 
-	nodeman := node.NewNodeManager()
-	require.NotNil(nodeman)
+	backend := api.NewStatusBackend()
```
Comment (adambabik) [inline geth/jail/jail_rpc_test.go]: You can make this test passing without getting `StatusBackend`. The only missing thing is to call 
```
txQueueManager.Start()
defer txQueueManager.Stop()
```

Getting the whole `StatusBackend` seems too heavy.

File: geth/node/txqueue_manager.go (hunk 17 above)
Comment (influx6) [reply geth/node/txqueue_manager.go]: Thanks, updated!

File: geth/jail/jail_rpc_test.go [hunk 21]
```diff
@@ -157,14 +173,48 @@ func (s *JailRPCTestSuite) TestSendTransaction() {
 	exported, err := result.Export()
 	require.NoError(err)
 
-	rawJSON, ok := exported.(json.RawMessage)
-	require.True(ok, "Expected raw json payload")
-	require.Equal(string(rawJSON), "\"3434=done\"")
+	fmt.Printf("Incoming response: %+q -> %+q\n", result, exported)
+
+	rawJSON, ok := exported.(string)
```
Comment (adambabik) [inline geth/jail/jail_rpc_test.go]: Maybe:
```
require.True(result.IsString())
require.NoEmpty(result.String())
```

File: geth/jail/execution_policy.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (tiabc) [reply geth/jail/execution_policy.go]: Nope :) Let's address this later.

File: geth/node/manager.go (hunk 15 above)
Comment (tiabc) [reply geth/node/manager.go]: I have an inner desire to remove all that as it's unrelated to what this function does but I'm not yet sure about the best way to accomplish it. Let's leave it for now.

File: geth/node/manager.go [hunk 22]
```diff
@@ -504,38 +504,22 @@ func (m *NodeManager) AccountKeyStore() (*keystore.KeyStore, error) {
...
-	}
-
-	m.RLock()
-	defer m.RUnlock()
+	m.Lock()
+	defer m.Unlock()
 
 	// make sure that node is fully started
-	if m.node == nil || m.nodeStarted == nil {
+	if m.node == nil {
```
Comment (tiabc) [inline geth/node/manager.go]: Why the second check has been removed?

File: geth/node/manager.go [hunk 23]
```diff
@@ -545,13 +529,53 @@ func (m *NodeManager) RPCClient() (*rpc.Client, error) {
...
 	}
 
+	return m.rpcClient, nil
+}
+
+// RPCUpstreamClient exposes reference to RPC client connected to the running node.
+func (m *NodeManager) RPCUpstreamClient() (*rpc.Client, error) {
+	if m == nil {
+		return nil, ErrInvalidNodeManager
+	}
```
Comment (tiabc) [inline geth/node/manager.go]: Guys, I think it's about time to stop mixing clear business logic with a lot of defensive mechanisms. These errors should be verified in tests.

File: geth/jail/jail_rpc_test.go [hunk 24]
```diff
@@ -137,8 +146,8 @@ func (s *JailRPCTestSuite) TestSendTransaction() {
 
 	// httpRPCServer will serve as an upstream server accepting transactions.
 	httpRPCServer := httptest.NewServer(rpcService)
-	s.StartTestNode(params.RopstenNetworkID, WithUpstream(httpRPCServer.URL))
-	defer s.StopTestNode()
+	s.StartTestBackend(params.RopstenNetworkID, WithUpstream(httpRPCServer.URL))
+	defer s.StopTestBackend()
```
[the remark this answers is not in the document]
Comment (adambabik) [reply geth/jail/jail_rpc_test.go]: This is a good remark. We already have a few tests that verify sending transactions through RPC:
1. https://github.com/status-im/status-go/blob/develop/geth/node/rpc_test.go#L68-L74
2. https://github.com/status-im/status-go/blob/develop/geth/api/backend_jail_test.go#L34

Does this test introduce anything more?

File: geth/node/manager.go (hunk 15 above)
Comment (influx6) [reply geth/node/manager.go]: Alright. I cant remember who, but someone did re-write the checks into a single method, am sure it is in one of the PRs.

File: geth/jail/jail_rpc_test.go (hunk 20 above)
Comment (influx6) [reply geth/jail/jail_rpc_test.go]: Ah! I see, thank you.

File: geth/jail/jail_rpc_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (influx6) [reply geth/jail/jail_rpc_test.go]: This tests were added to validate sending transactions to the upstream RPC, do we consider the tests already present good enough and this un-necessary? @adambabik

File: geth/jail/jail_rpc_test.go (hunk 19 above)
Comment (influx6) [reply geth/jail/jail_rpc_test.go]: @adambabik I was considering the fact that we would need to add this piece into the whole approach of sending transactions using the upstream node and from the looks of things, this is not done automatically and must be done for calls when we wish to complete a transaction, like we did here. 

- Is there a way to baked this part as part of the send transaction?
- Are there already a mechanism in place that appopriately calls `CompleteTransaction` with necessary `ChatID` and `password` when running in production?

File: geth/node/manager.go (hunk 15 above)
Comment (adambabik) [reply geth/node/manager.go]: I took over that PR but it might not get merged as I tested the issue with the current develop branch and everything worked as expected, meaning it looks like no changes are needed.

File: geth/jail/jail_rpc_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (adambabik) [reply geth/jail/jail_rpc_test.go]: Oh, no. These old tests I mentioned probably do not test the upstream RPC client. But maybe we can put this test in there to keep them in the same place and be consistent?

File: geth/jail/execution_policy.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (influx6) [reply geth/jail/execution_policy.go]: Ah! Alright :(

File: geth/jail/jail_rpc_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (influx6) [reply geth/jail/jail_rpc_test.go]: Sure, we can move so, should I do so in this PR? @adambabik

File: geth/jail/jail_rpc_test.go (hunk 19 above)
Comment (influx6) [reply geth/jail/jail_rpc_test.go]: @adambabik Your thoughts?

File: geth/node/manager.go (hunk 15 above)
Comment (influx6) [reply geth/node/manager.go]: Ok. We can discuss this a little then refactor in another issue, as @tiabc has thoughts on this.

File: geth/jail/jail_rpc_test.go (hunk 19 above)
Comment (adambabik) [reply geth/jail/jail_rpc_test.go]: But sending the transaction using the upstream is still two-step process and it works the same way. First, one triggers "send transaction" (it can be done via a C binding, RPC call, `jail.Send` that is bound to `send` function) but it merely puts the transaction into the queue. Next, it must be completed and only then the transaction is actually sent. Again, it works the same way with the upstream. 

Yes, `CompleteTransaction` is provided as a C binding and works in the current develop. With the upstream it should work the same way, just call a proper RPC client. Check out `TxQueueManager.completeRemoteTransaction`. I think you created it in `ExecutionPolicy`, I just copied it.

It can be called manually, however. Check out the following changes:

```
@@ -42,8 +43,9 @@ func TestJailRPCTestSuite(t *testing.T) {

 type JailRPCTestSuite struct {
        BaseTestSuite
-       Account common.AccountManager
-       Policy  *jail.ExecutionPolicy
+       Account        common.AccountManager
+       TxQueueManager common.TxQueueManager
+       Policy         *jail.ExecutionPolicy
 }

@@ -56,6 +58,11 @@ func (s *JailRPCTestSuite) SetupTest() {
        require.NotNil(acctman)

        txQueueManager := node.NewTxQueueManager(nodeman, acctman)
+       txQueueManager.SetTransactionQueueHandler(func(queuedTx *common.QueuedTx) {
+               s.T().Logf("queued a new transaction: %s", queuedTx.ID)
+               _, err := txQueueManager.CompleteTransaction(queuedTx.ID, TestConfig.Account1.Password)
+               s.NoError(err)
+       })

        policy := jail.NewExecutionPolicy(nodeman, acctman, txQueueManager)
        require.NotNil(policy)
@@ -63,6 +70,7 @@ func (s *JailRPCTestSuite) SetupTest() {
        s.Policy = policy
        s.Account = acctman
        s.NodeManager = nodeman
+       s.TxQueueManager = txQueueManager
 }

        // httpRPCServer will serve as an upstream server accepting transactions.
        httpRPCServer := httptest.NewServer(rpcService)
        s.StartTestNode(params.RopstenNetworkID, WithUpstream(httpRPCServer.URL))
        defer s.StopTestNode()

+       s.TxQueueManager.Start()
+       defer s.TxQueueManager.Stop()
```

Comment (influx6) [reply geth/jail/jail_rpc_test.go]: Alright thanks.

File: geth/jail/jail_rpc_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (adambabik) [reply geth/jail/jail_rpc_test.go]: We can test it here, following https://github.com/status-im/status-go/pull/276#discussion_r137028856, but I feel like it would be better to move this test to `rpc_test.go`. Then, we could have all RPC tests in one place.

File: geth/node/manager.go [hunk 25]
```diff
@@ -54,10 +54,6 @@ func NewNodeManager() *NodeManager {
 
 // StartNode start Status node, fails if node is already started
 func (m *NodeManager) StartNode(config *params.NodeConfig) (<-chan struct{}, error) {
-	if m == nil {
```
Comment (adambabik) [inline geth/node/manager.go]: Hm, why these changes appeared in this PR?

File: geth/node/rpc_test.go [hunk 26]
```diff
@@ -33,7 +57,121 @@ func (s *RPCTestSuite) SetupTest() {
...
+	policy := jail.NewExecutionPolicy(nodeManager, acctman, txQueueManager)
+	require.NotNil(policy)
+
+	s.Policy = policy
+	s.Account = acctman
 	s.NodeManager = nodeManager
+	s.TxQueueManager = txQueueManager
+}
+
+func (s *RPCTestSuite) TestRPCSendTransaction() {
```
Comment (adambabik) [inline geth/node/rpc_test.go]: By moving I also meant to make it similar to `TestCallRPC` :P So, I would get rid of all these implementation details like `txQueueManager`, `ExecutionPolicy` and just follow the pattern:
1. Get `rpcClient`,
2. Generate config with upstream enabled,
3. Start node,
4. Call `rpcClient.Call(inputJSON)`.

There is an example with `eth_sendTransaction` in `TestCallRPC`.

File: geth/node/manager.go (hunk 25 above)
Comment (influx6) [reply geth/node/manager.go]: Since @tiabc mention our need to stop checking such logic which should rather be validated in tests, felt it best to remove this, is there a problem? Do we wish to move it into a PR of it's own?

File: geth/node/rpc_test.go (hunk 26 above)
Comment (influx6) [reply geth/node/rpc_test.go]: Ah! I see

File: geth/node/manager.go (hunk 25 above)
Comment (adambabik) [reply geth/node/manager.go]: Up to @tiabc, but I saw a few of his comments when he insisted on not introducing refactoring changes into feature PRs.

File: geth/node/rpc_test.go [hunk 27]
```diff
@@ -32,10 +47,69 @@ func (s *RPCTestSuite) SetupTest() {
...
+	require := s.Require()
+	expectedResponse := []byte(`{"jsonrpc": "2.0", "status":200, "result": "3434=done"}`)
+
+	// httpRPCServer will serve as an upstream server accepting transactions.
+	httpRPCServer := httptest.NewServer(service{
+		Handler: func(w http.ResponseWriter, r *http.Request) {
+			defer r.Body.Close()
+
+			var txReq txRequest
+
```
Comment (adambabik) [inline geth/node/rpc_test.go]: Minor: this space is not needed.

File: geth/node/rpc_test.go [hunk 28]
```diff
@@ -32,10 +47,69 @@ func (s *RPCTestSuite) SetupTest() {
...
+			err := json.NewDecoder(r.Body).Decode(&txReq)
+			require.NoError(err)
+
+			switch txReq.Method {
+			case "eth_getTransactionCount":
```
Comment (adambabik) [inline geth/node/rpc_test.go]: Are there any other possible `txReq.Method` values in this test that we care about? If not maybe use `if`?

File: geth/node/rpc_test.go [hunk 29]
```diff
@@ -32,10 +47,69 @@ func (s *RPCTestSuite) SetupTest() {
...
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte(`{"jsonrpc": "2.0", "status":200, "result": "0x434"}`))
+				return
+			}
+
+			payload := ([]byte)(txReq.Payload)
+
+			var bu []interface{}
+
```
Comment (adambabik) [inline geth/node/rpc_test.go]: Unnecessary blank line.

File: geth/node/rpc_test.go [hunk 30]
```diff
@@ -32,10 +47,69 @@ func (s *RPCTestSuite) SetupTest() {
...
+			jserr := json.Unmarshal(payload, &bu)
+			require.NoError(jserr)
+			require.NotNil(bu)
```
Comment (adambabik) [inline geth/node/rpc_test.go]: Not needed are there is `require.Len` below.

File: geth/node/manager.go (hunk 25 above)
Comment (tiabc) [reply geth/node/manager.go]: Yeah, if you just noticed something in unrelated part of code, just let it go and add to a refactoring note. The goal here is to ease merging.
In this case they're unrelated and should have been skipped but we're not going to discuss it and waste time, let's just merge it as soon as possible and keep it in mind next time.

File: geth/node/rpc_test.go [hunk 31]
```diff
@@ -11,6 +13,17 @@ import (
 	"github.com/stretchr/testify/suite"
 )
 
+//==================================================================================================
```
Comment (tiabc) [inline geth/node/rpc_test.go]: I think we should stop using separators, I don't know even a single case where they'd increase readability. Their usage often implies that it's time to separate entities or functionality over several files.

File: geth/node/manager.go [hunk 32]
```diff
@@ -545,19 +473,47 @@ func (m *NodeManager) RPCClient() (*rpc.Client, error) {
...
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
 	if m.rpcClient == nil {
-		return nil, ErrInvalidRPCClient
+		m.rpcClient, err = rpc.Dial(config.UpstreamConfig.URL)
+		if err != nil {
+			log.Error("Failed to conect to upstream RPC server", "error", err)
```
Comment (rgeraldes) [inline geth/node/manager.go]: 'connect'

File: geth/node/manager.go (hunk 25 above)
Comment (influx6) [reply geth/node/manager.go]: Alright!
